src/sinks/file/mod.rs

```diff
@@ -88,98 +95,143 @@ impl PartitionedFileSink {
 
         Some(bytes)
     }
-}
-
-impl Sink for PartitionedFileSink {
-    type SinkItem = Event;
-    type SinkError = ();
-
-    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        if let Some(key) = self.partition_event(&event) {
-            let encoding = self.encoding.clone();
 
-            let partition = self
-                .partitions
-                .entry(key.clone())
-                .or_insert_with(|| File::new(key.clone(), encoding.clone()));
-
-            self.last_accessed.insert(key.clone(), Instant::now());
+    fn deadline_at(&self) -> Instant {
+        Instant::now()
+            .checked_add(self.idle_timeout)
+            .expect("unable to compute next deadline")
+    }
 
-            partition
-                .start_send(event)
-                .map_err(|error| error!(message = "Error writing to partition.", %error))
-        } else {
-            Ok(AsyncSink::Ready)
+    async fn run(&mut self, input: impl Stream<Item = Event> + Send + Sync) -> crate::Result<()> {
+        pin_mut!(input);
+        loop {
+            tokio02::select! {
+                event = input.next() => {
+                    match event {
+                        None => {
+                            // If we got `None` - terminate the processing.
+                            debug!(message = "Receiver exausted, terminating the processing loop.");
+                            break;
+                        }
+                        Some(event) => self.process_event(event).await,
+                    }
+                }
+                result = self.files.next() => {
```

Comment:
I can't remember but can we match on the left side?

```
Some(Ok(..)) = self.files.next() => ...,
Some(Err(...)) = self.files.next() => ...,
```
etc?

Reply:
The syntax, just from the looks of it, should be causing a multi-mut error. I can try, but it surely looks cuter intuitive, and I doubt it'll work.

Reply:
It's now allowed. `first mutable borrow occurs here`, `second mutable borrow occurs here` - this kind of errors.